# Animal Control Data Preparation
## Rebecca Lewis
## DSC 680

In [1]:
import pandas as pd
import numpy as np
import datetime

### Review the cleaned dataset and the raw dataset and merge them based on file_number.

In [2]:
# load pre cleaned and raw data
ac_clean = pd.read_csv('../Data/BR_Animal_Control_Calls.csv')
ac_raw = pd.read_csv('../Data/RAW_Animal_Control_Incidents.csv')

ac_clean.head(), ac_raw.head()

C:\Users\rvick\Anaconda3\envs\R_Python\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (14,32) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(   file_number         incident_date          request_type  \
 0    201639027  2016-12-27T00:00:00Z          CRUELTY CASE   
 1    201639044  2016-12-28T00:00:00Z        TRAPPED ANIMAL   
 2    201639036  2016-12-28T00:00:00Z  STRAY ANIMAL SECURED   
 3    201639042  2016-12-28T00:00:00Z         ABANDONED DOG   
 4    201639059  2016-12-28T00:00:00Z  STRAY ANIMAL SECURED   
 
                                 location municipality  zip_code   species  \
 0        3366 SYCAMORE, BATON ROUGE, LA.  BATON ROUGE   70805.0       DOG   
 1  2582 WOODLAND RIDGE, BATON ROUGE, LA.  BATON ROUGE   70816.0  WILDLIFE   
 2                4463 MAIN, ZACHARY, LA.      ZACHARY   70791.0       DOG   
 3            2100 ERIE, BATON ROUGE, LA.  BATON ROUGE   70805.0       DOG   
 4        299 VAN BUREN, BATON ROUGE, LA.  BATON ROUGE   70714.0       DOG   
 
                breed sex    size  age  condition temperment     disposition  \
 0            PITBULL   F   LARGE  NaN       FAIR   FRIENDLY       TRA

In [3]:
# get the size of data
ac_clean.shape, ac_raw.shape

((52585, 16), (95329, 38))

In [4]:
ac_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52585 entries, 0 to 52584
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   file_number    52585 non-null  int64  
 1   incident_date  52585 non-null  object 
 2   request_type   19735 non-null  object 
 3   location       52585 non-null  object 
 4   municipality   52585 non-null  object 
 5   zip_code       24185 non-null  float64
 6   species        52564 non-null  object 
 7   breed          51770 non-null  object 
 8   sex            40237 non-null  object 
 9   size           51315 non-null  object 
 10  age            32906 non-null  float64
 11  condition      50383 non-null  object 
 12  temperment     50986 non-null  object 
 13  disposition    52578 non-null  object 
 14  lat            52585 non-null  float64
 15  long           52585 non-null  float64
dtypes: float64(4), int64(1), object(11)
memory usage: 6.4+ MB


In [5]:
ac_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95329 entries, 0 to 95328
Data columns (total 38 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   FILE NUMBER           95329 non-null  int64  
 1   INCIDENT DATE         95329 non-null  object 
 2   INCIDENT TIME         94844 non-null  object 
 3   DISPATCHED TIME       84573 non-null  object 
 4   CALL TAKER            75160 non-null  object 
 5   DISPATCHER            82265 non-null  object 
 6   DISPATCHED OTHER      95329 non-null  int64  
 7   DISPATCHED SITUATION  84563 non-null  object 
 8   ARRIVAL TIME          83869 non-null  object 
 9   AVAILABLE TIME        84571 non-null  object 
 10  COMPLAINANT STREET    95273 non-null  object 
 11  REQUEST TYPE          62167 non-null  object 
 12  OFFICER               83261 non-null  object 
 13  EQUIPMENT NUMBER      80390 non-null  object 
 14  REMARKS               2098 non-null   object 
 15  IMPOUND NUMBER     

In [6]:
# merge based on file number with inner join from the clean table.

ac_full = pd.merge(ac_clean, ac_raw, left_on='file_number', right_on='FILE NUMBER', how='inner')
ac_full.head()

file_number         incident_date          request_type  \
0    201639027  2016-12-27T00:00:00Z          CRUELTY CASE   
1    201639044  2016-12-28T00:00:00Z        TRAPPED ANIMAL   
2    201639036  2016-12-28T00:00:00Z  STRAY ANIMAL SECURED   
3    201639042  2016-12-28T00:00:00Z         ABANDONED DOG   
4    201639059  2016-12-28T00:00:00Z  STRAY ANIMAL SECURED   

                                location municipality  zip_code   species  \
0        3366 SYCAMORE, BATON ROUGE, LA.  BATON ROUGE   70805.0       DOG   
1  2582 WOODLAND RIDGE, BATON ROUGE, LA.  BATON ROUGE   70816.0  WILDLIFE   
2                4463 MAIN, ZACHARY, LA.      ZACHARY   70791.0       DOG   
3            2100 ERIE, BATON ROUGE, LA.  BATON ROUGE   70805.0       DOG   
4        299 VAN BUREN, BATON ROUGE, LA.  BATON ROUGE   70714.0       DOG   

               breed sex    size  ...  CONDITION TEMPERMENT   COLLAR  \
0            PITBULL   F   LARGE  ...       FAIR   FRIENDLY    NYLON   
1             POSSUM   U   LARGE  ...       FAIR     NORMAL  UNKNOWN   
2            PITBULL   M   LARGE  ...  EMACIATED   FRIENDLY  LEATHER   
3    GERMAN SHEPHERD   F  MEDIUM  ...  EMACIATED   FRIENDLY  UNKNOWN   
4  YORKSHIRE TERRIER   F   SMALL  ...  EMACIATED     NORMAL  UNKNOWN   

  COLLAR COLOR  VACCINATION NUMBER  VACCINATION DATE  PET NAME  \
0          RED                 NaN               NaN       NaN   
1      UNKNOWN                 NaN               NaN       NaN   
2         PINK                 NaN               NaN       NaN   
3      UNKNOWN                 NaN               NaN       NaN   
4      UNKNOWN                 NaN               NaN       NaN   

      DISPOSITION DISPOSITION DATE DISPOSITION OFFICER  
0       TRANS CAA       12/27/2016             FRANK S  
1  RETURN TO WILD       12/28/2016            TRAVIS R  
2       TRANS CAA       12/28/2016             ISAAC S  
3       TRANS CAA       12/28/2016              JOHN B  
4       TRANS CAA       12/28/2016               SUE H  

[5 rows x 54 columns]

In [7]:
#check shape
ac_full.shape

(102635, 54)

In [8]:
# there seems to be a cartesian product.  Checking for duplicate file numbers in each file
ac_clean.groupby(['file_number']).count()['incident_date'].nlargest(1000)

file_number
2014274530    72
201862120     46
2012232690    45
201745155     30
201752463     28
              ..
201739440      3
201739661      3
201739804      3
201739805      3
201739882      3
Name: incident_date, Length: 1000, dtype: int64

In [9]:
#view a specific example
ac_clean[ac_clean['file_number'] == 201739804] 

file_number         incident_date          request_type  \
41662    201739804  2017-01-19T00:00:00Z  STRAY ANIMAL SECURED   
41664    201739804  2017-01-19T00:00:00Z  STRAY ANIMAL SECURED   
41667    201739804  2017-01-19T00:00:00Z  STRAY ANIMAL SECURED   

                                 location municipality  zip_code species  \
41662  15208 OLD SCENIC, BAKER, LA. 70714        BAKER   70714.0     CAT   
41664  15208 OLD SCENIC, BAKER, LA. 70714        BAKER   70714.0     CAT   
41667     15208 OLD SCENIC, BAKER, LA. NA        BAKER       NaN     CAT   

      breed sex     size  age condition temperment disposition        lat  \
41662   DSH   U  UNKNOWN  NaN      FAIR     NORMAL   TRANS CAA  30.590448   
41664   DMH   U  UNKNOWN  NaN   UNKNOWN     NORMAL   TRANS CAA  30.590448   
41667   DMH   U    LARGE  NaN      FAIR   FRIENDLY   TRANS CAA  30.590448   

            long  
41662 -91.205236  
41664 -91.205236  
41667 -91.205236

### Problem
It looks like if multiple animals are picked up at an incident, there are more than one observation for the file number, this will make merging the datasets difficult.  There are no other identifying facters that would be considered unique so joining the datasets on the file number will not work.  To combat this, I can extract the latitude and longitude to a separate dataset with the file number and location to reference from the raw dataset.  

In [10]:
# extract geocoded location information
ac_geo_info = ac_clean[['file_number', 'location', 'zip_code','lat','long']]
ac_geo_info.head(), ac_geo_info.shape

(   file_number                               location        lat       long
 0    201639027        3366 SYCAMORE, BATON ROUGE, LA.  30.494640 -91.156160
 1    201639044  2582 WOODLAND RIDGE, BATON ROUGE, LA.  30.431417 -91.141070
 2    201639036                4463 MAIN, ZACHARY, LA.  30.648540 -91.156770
 3    201639042            2100 ERIE, BATON ROUGE, LA.  30.476015 -91.168460
 4    201639059        299 VAN BUREN, BATON ROUGE, LA.  30.429104 -91.185036,
 (52585, 4))

In [11]:
# drop duplicates and verify shape 
ac_geo_info = ac_geo_info.drop_duplicates()
ac_geo_info.shape 

(44164, 4)

In [12]:
#save the geocoded references
ac_geo_info.to_csv('../Data/geocoded_locations.csv', header=True, index=False)

###  View statistics on columns to determine additional cleanup 

In [13]:
ac_raw.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95329 entries, 0 to 95328
Data columns (total 38 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   FILE NUMBER           95329 non-null  int64  
 1   INCIDENT DATE         95329 non-null  object 
 2   INCIDENT TIME         94844 non-null  object 
 3   DISPATCHED TIME       84573 non-null  object 
 4   CALL TAKER            75160 non-null  object 
 5   DISPATCHER            82265 non-null  object 
 6   DISPATCHED OTHER      95329 non-null  int64  
 7   DISPATCHED SITUATION  84563 non-null  object 
 8   ARRIVAL TIME          83869 non-null  object 
 9   AVAILABLE TIME        84571 non-null  object 
 10  COMPLAINANT STREET    95273 non-null  object 
 11  REQUEST TYPE          62167 non-null  object 
 12  OFFICER               83261 non-null  object 
 13  EQUIPMENT NUMBER      80390 non-null  object 
 14  REMARKS               2098 non-null   object 
 15  IMPOUND NUMBER     

In [14]:
#rename columns to remove spaces and uppercase to make it easier to work with
new_columns = ac_raw.columns
new_columns = list(map(str.lower, new_columns.str.replace(' ','_')))
ac_raw.columns = new_columns
ac_raw.head()

file_number incident_date incident_time dispatched_time call_taker  \
0    201639027    12/27/2016         15:30           16:06    UNKNOWN   
1    201639044    12/28/2016         10:06           10:07    UNKNOWN   
2    201639036    12/28/2016         09:03           09:03    UNKNOWN   
3    202087899    08/30/2020         13:39           00:00    UNKNOWN   
4    201639059    12/28/2016         15:05           15:45    UNKNOWN   

   dispatcher  dispatched_other dispatched_situation arrival_time  \
0  JENNIFER G                 0                    D        16:17   
1  JENNIFER G                 0                    D        10:15   
2  JENNIFER G                 0                    D        09:27   
3     UNKNOWN                 0                    D        00:00   
4  JENNIFER G                 0                    D        16:07   

  available_time  ...  condition temperment   collar collar_color  \
0          16:38  ...       FAIR   FRIENDLY    NYLON          RED   
1          10:23  ...       FAIR     NORMAL  UNKNOWN      UNKNOWN   
2          09:45  ...  EMACIATED   FRIENDLY  LEATHER         PINK   
3          16:52  ...        NaN        NaN      NaN          NaN   
4          16:25  ...  EMACIATED     NORMAL  UNKNOWN      UNKNOWN   

  vaccination_number  vaccination_date pet_name     disposition  \
0                NaN               NaN      NaN       TRANS CAA   
1                NaN               NaN      NaN  RETURN TO WILD   
2                NaN               NaN      NaN       TRANS CAA   
3                NaN               NaN      NaN             NaN   
4                NaN               NaN      NaN       TRANS CAA   

  disposition_date disposition_officer  
0       12/27/2016             FRANK S  
1       12/28/2016            TRAVIS R  
2       12/28/2016             ISAAC S  
3              NaN                 NaN  
4       12/28/2016               SUE H  

[5 rows x 38 columns]

In [15]:
# view summary statistics for each column
ac_summary = ac_raw.describe(include=np.object).T
ac_summary

count unique                     top   freq
incident_date         95329   2877              07/17/2017    133
incident_time         94844   1622                   08:11    438
dispatched_time       84573   1091                   00:00   6602
call_taker            75160     74                 UNKNOWN  49057
dispatcher            82265     21              JENNIFER G  46445
dispatched_situation  84563      5                       D  78108
arrival_time          83869   1078                   00:00   8699
available_time        84571   1090                   00:00   6140
complainant_street_   95273   7858    GREENWELL SPRINGS RD    991
request_type          62167     49                 UNKNOWN   9148
officer               83261     80                 UNKNOWN  19890
equipment_number      80390     69                    AC04   9107
remarks                2098    977                     RTW    210
impound_date          53194   2812              05/25/2014     73
impound_time          53135   1731                   10:30    278
location              52956  27500           2680 PROGRESS    296
municipality_         53195      6             BATON ROUGE  48268
service_code          53199     27          TRAPPED ANIMAL  17667
species               53177     11                     DOG  26898
breed                 52371    186                     DSH  11698
sex                   40731      9                       M  14395
size                  51904      5                  MEDIUM  23415
color                 51843   2210                     BLK   4089
condition             50967      6                    FAIR  34897
temperment            51575      6                  NORMAL  24684
collar                25055    110                 UNKNOWN  16069
collar_color          24538     17                 UNKNOWN  16819
vaccination_number      600    573                    NONE      6
vaccination_date        348    310  09/03/2014 12:00:00 AM      6
pet_name               1014    635                    COCO     17
disposition           53192      7               TRANS CAA  39834
disposition_date      38610   2220              05/25/2014     73
disposition_officer   52412     74                     SW1  19708

In [16]:
ac_raw.describe().T

count          mean           std          min  \
file_number       95329.0  8.321673e+08  8.630312e+08  201629133.0   
dispatched_other  95329.0  1.560910e-02  1.239581e-01          0.0   
impound_number    53199.0  6.092900e+04  8.762242e+05          0.0   
zip_code          24310.0  7.080073e+04  2.191823e+02      55887.0   
age               33217.0  2.318090e-02  4.350987e-01          0.0   

                          25%          50%           75%           max  
file_number       201751156.0  201973420.0  2.013248e+09  2.016307e+09  
dispatched_other          0.0          0.0  0.000000e+00  1.000000e+00  
impound_number        37425.5      55076.0  7.558750e+04  2.018633e+08  
zip_code              70803.0      70807.0  7.081400e+04  7.801400e+04  
age                       0.0          0.0  0.000000e+00  1.700000e+01

In [17]:
# account for missing time values - causing an error when converting to date time format
ac_raw = ac_raw.replace(':', '00:00')
ac_raw[ac_raw['dispatched_time'] == ':']

Empty DataFrame
Columns: [file_number, incident_date, incident_time, dispatched_time, call_taker, dispatcher, dispatched_other, dispatched_situation, arrival_time, available_time, complainant_street_, request_type, officer, equipment_number, remarks, impound_number, impound_date, impound_time, location, municipality_, zip_code, service_code, species, breed, sex, size, color, age, condition, temperment, collar, collar_color, vaccination_number, vaccination_date, pet_name, disposition, disposition_date, disposition_officer]
Index: []

[0 rows x 38 columns]

In [18]:
#combine dates and times for data points to make them easier to work with

ac_raw['incident_datetime'] = ac_raw['incident_date'].str.cat(ac_raw['incident_time'], sep =" ")
ac_raw['dispatched_datetime'] = ac_raw['incident_date'].str.cat(ac_raw['dispatched_time'], sep =" ")
ac_raw['arrival_datetime'] = ac_raw['incident_date'].str.cat(ac_raw['arrival_time'], sep =" ")
ac_raw['available_datetime'] = ac_raw['incident_date'].str.cat(ac_raw['available_time'], sep =" ")
ac_raw['impound_datetime'] = ac_raw['impound_date'].str.cat(ac_raw['impound_time'], sep =" ")

In [19]:
ac_raw['incident_datetime'].head()

0    12/27/2016 15:30
1    12/28/2016 10:06
2    12/28/2016 09:03
3    08/30/2020 13:39
4    12/28/2016 15:05
Name: incident_datetime, dtype: object

In [20]:
# convert datetimes: incident_date, incident_time, dispatched_time, arrival_time, available_time, impound_date, impound_time, vaccination_date, disposition_date
# get min and max date times
ac_raw['incident_datetime'] = ac_raw['incident_datetime'].apply(pd.to_datetime, format='%m/%d/%Y %H:%M')
ac_raw['dispatched_datetime'] = ac_raw['dispatched_datetime'].apply(pd.to_datetime, format='%m/%d/%Y %H:%M')
ac_raw['arrival_datetime'] = ac_raw['arrival_datetime'].apply(pd.to_datetime, format='%m/%d/%Y %H:%M')
ac_raw['available_datetime'] = ac_raw['available_datetime'].apply(pd.to_datetime, format='%m/%d/%Y %H:%M', errors='coerce')
ac_raw['impound_datetime']=ac_raw['impound_datetime'].apply(pd.to_datetime, format='%m/%d/%Y %H:%M', errors='coerce')
ac_raw['disposition_date'] = ac_raw['disposition_date'].apply(pd.to_datetime, format='%m/%d/%Y', errors='coerce')
ac_raw['vaccination_date'] = ac_raw['vaccination_date'].apply(pd.to_datetime, format='%m/%d/%Y %H:%M:%S %p', errors='coerce')

In [21]:
# get min and max dates
ac_raw[['incident_datetime', 'dispatched_datetime','arrival_datetime', 'available_datetime', 'impound_datetime', 'disposition_date',
'vaccination_date']].agg(("min", "max"))

incident_datetime dispatched_datetime    arrival_datetime  \
min 2012-01-01 11:20:00 2012-01-03 08:19:00 2012-01-03 08:52:00   
max 2020-11-18 08:00:00 2020-11-18 00:00:00 2020-11-18 00:00:00   

     available_datetime    impound_datetime disposition_date  \
min 2012-01-03 08:55:00 1951-10-12 19:20:00       1951-10-12   
max 2020-11-18 16:09:00 2020-07-28 15:50:00       2020-06-27   

       vaccination_date  
min 2009-07-20 12:00:00  
max 2019-08-06 12:00:00

In [22]:
#convert impound number to string

ac_raw['impound_number'] = np.nan_to_num(ac_raw['impound_number']).astype(int)
ac_raw['zip_code'] = np.nan_to_num(ac_raw['zip_code']).astype(int)

### Column Elimination
The COMPLAINANT STREET was removed to protect the identity of the complainant.

These columns were removed because they are not likely to be relevant to the analysis:
- CALL TAKER
- DISPATCHER
- DISPATCHED OTHER
- OFFICER
- EQUIPMENT NUMBER
- COLLAR COLOR
- DISPOSITION OFFICER

Compared to the amount of data in the dataset, these variables are not likely to have enough occurences to provide meaningful information:
- VACCINATION NUMBER
- VACCINATION DATE
- PET NAME

In [23]:
# drop unnecessary columns
ac_new_clean = ac_raw.drop(['complainant_street_','call_taker', 'dispatcher', 'dispatched_other', 'officer', 'equipment_number', 'collar_color','disposition_officer', 'vaccination_number', 'vaccination_date', 'pet_name'], axis=1)
ac_new_clean.head()

file_number incident_date incident_time dispatched_time  \
0    201639027    12/27/2016         15:30           16:06   
1    201639044    12/28/2016         10:06           10:07   
2    201639036    12/28/2016         09:03           09:03   
3    202087899    08/30/2020         13:39           00:00   
4    201639059    12/28/2016         15:05           15:45   

  dispatched_situation arrival_time available_time          request_type  \
0                    D        16:17          16:38          CRUELTY CASE   
1                    D        10:15          10:23        TRAPPED ANIMAL   
2                    D        09:27          09:45  STRAY ANIMAL SECURED   
3                    D        00:00          16:52               UNKNOWN   
4                    D        16:07          16:25  STRAY ANIMAL SECURED   

  remarks  impound_number  ...  condition temperment   collar     disposition  \
0     NaN           72536  ...       FAIR   FRIENDLY    NYLON       TRANS CAA   
1     NaN           72414  ...       FAIR     NORMAL  UNKNOWN  RETURN TO WILD   
2     NaN           70587  ...  EMACIATED   FRIENDLY  LEATHER       TRANS CAA   
3     NaN               0  ...        NaN        NaN      NaN             NaN   
4     NaN           71349  ...  EMACIATED     NORMAL  UNKNOWN       TRANS CAA   

   disposition_date   incident_datetime dispatched_datetime  \
0        2016-12-27 2016-12-27 15:30:00 2016-12-27 16:06:00   
1        2016-12-28 2016-12-28 10:06:00 2016-12-28 10:07:00   
2        2016-12-28 2016-12-28 09:03:00 2016-12-28 09:03:00   
3               NaT 2020-08-30 13:39:00 2020-08-30 00:00:00   
4        2016-12-28 2016-12-28 15:05:00 2016-12-28 15:45:00   

     arrival_datetime  available_datetime    impound_datetime  
0 2016-12-27 16:17:00 2016-12-27 16:38:00 2016-12-27 16:20:00  
1 2016-12-28 10:15:00 2016-12-28 10:23:00 2016-12-28 10:21:00  
2 2016-12-28 09:27:00 2016-12-28 09:45:00 2016-12-28 09:41:00  
3 2020-08-30 00:00:00 2020-08-30 16:52:00                 NaT  
4 2016-12-28 16:07:00 2016-12-28 16:25:00 2016-12-28 16:20:00  

[5 rows x 32 columns]

In [24]:
ac_new_clean.shape

(95329, 32)

In [25]:
# add lat and long columns from lookup file
ac_new_clean = pd.merge(ac_new_clean, ac_geo_info, left_on='file_number', right_on='file_number', how='inner')

By inner joining on the previously cleaned data's geocoded records, I'm eliminating rows that were previously identified as not being relevant for analysis and still taking advantage of previously cleansed data.

In [26]:
ac_new_clean.shape

(55889, 35)

In [27]:
ac_new_clean.head()

file_number incident_date incident_time dispatched_time  \
0    201639027    12/27/2016         15:30           16:06   
1    201639044    12/28/2016         10:06           10:07   
2    201639036    12/28/2016         09:03           09:03   
3    201639059    12/28/2016         15:05           15:45   
4    201639043    12/28/2016         10:03           10:10   

  dispatched_situation arrival_time available_time          request_type  \
0                    D        16:17          16:38          CRUELTY CASE   
1                    D        10:15          10:23        TRAPPED ANIMAL   
2                    D        09:27          09:45  STRAY ANIMAL SECURED   
3                    D        16:07          16:25  STRAY ANIMAL SECURED   
4                    D        10:35          10:52  OWNER SIGNED RELEASE   

  remarks  impound_number  ...     disposition disposition_date  \
0     NaN           72536  ...       TRANS CAA       2016-12-27   
1     NaN           72414  ...  RETURN TO WILD       2016-12-28   
2     NaN           70587  ...       TRANS CAA       2016-12-28   
3     NaN           71349  ...       TRANS CAA       2016-12-28   
4     NaN           70589  ...       TRANS CAA       2016-12-28   

    incident_datetime dispatched_datetime    arrival_datetime  \
0 2016-12-27 15:30:00 2016-12-27 16:06:00 2016-12-27 16:17:00   
1 2016-12-28 10:06:00 2016-12-28 10:07:00 2016-12-28 10:15:00   
2 2016-12-28 09:03:00 2016-12-28 09:03:00 2016-12-28 09:27:00   
3 2016-12-28 15:05:00 2016-12-28 15:45:00 2016-12-28 16:07:00   
4 2016-12-28 10:03:00 2016-12-28 10:10:00 2016-12-28 10:35:00   

   available_datetime    impound_datetime  \
0 2016-12-27 16:38:00 2016-12-27 16:20:00   
1 2016-12-28 10:23:00 2016-12-28 10:21:00   
2 2016-12-28 09:45:00 2016-12-28 09:41:00   
3 2016-12-28 16:25:00 2016-12-28 16:20:00   
4 2016-12-28 10:52:00 2016-12-28 10:41:00   

                              location_y        lat       long  
0        3366 SYCAMORE, BATON ROUGE, LA.  30.494640 -91.156160  
1  2582 WOODLAND RIDGE, BATON ROUGE, LA.  30.431417 -91.141070  
2                4463 MAIN, ZACHARY, LA.  30.648540 -91.156770  
3        299 VAN BUREN, BATON ROUGE, LA.  30.429104 -91.185036  
4      6621 MARIONETTE, BATON ROUGE, LA.  30.529593 -91.125534  

[5 rows x 35 columns]

### Evaluate the distribution of categorical values

In [28]:
#categorical columns to review:['dispatched_situation', 'request_type', 'municipality_', 'zip_code', 'service_code', 'species', 
#               'breed', 'sex', 'size', 'color', 'condition', 'temperment', 'collar', 'disposition']

ac_new_clean['dispatched_situation'].value_counts(dropna=False)


D      42923
NaN    11654
K       1234
U         66
F         11
N          1
Name: dispatched_situation, dtype: int64

I'm unable to find what the values of U, F or N mean.  D means dispatched immediately, K means dispatched the following day

In [29]:
ac_new_clean['request_type'].value_counts(dropna=False)

NaN                           35157
TRAPPED ANIMAL                 5847
STRAY ANIMAL SECURED           2894
STRAY ANIMAL                   2227
UNKNOWN                        2216
INJURED ANIMAL                  684
ABANDONED ANIMAL                676
LOOSE ANIMAL                    599
OWNER SIGNED RELEASE            599
PITBULL                         529
UNIT STANDING BY                526
BITE CASE                       485
CRUELTY CASE                    445
VET PICKUP                      441
ANIMAL RESCUE                   427
DANGEROUS ANIMAL SITUATION      370
WILD LIFE                       367
SICK ANIMAL                     224
SPECIAL CASE                    221
OTHER                           185
SNAKE                           130
SCHOOL CALL                     130
CRUELTY RECHECK                 119
LOOSE LIVE STOCK                112
DELIVER TRAP                     99
PATROL AREA                      64
PICK UP TRAP                     57
DOG FIGHTING                

There are a large number of null values. Will need to determine how to handle.  The following categories can be combined:
STRAY/STRAY ANIMAL
CRUELTY RECHECK/RECHECK

In [30]:
ac_new_clean['municipality_'].value_counts(dropna=False)

BATON ROUGE    50724
PARISH          1763
ZACHARY         1289
BAKER           1046
CENTRAL          923
UNKNOWN          143
NaN                1
Name: municipality_, dtype: int64

This data is heavily skewed towards baton rouge.

In [31]:
ac_new_clean['zip_code'].value_counts(dropna=False)

0        30491
70805     3385
70816     1998
70802     1996
70815     1983
         ...  
70708        1
70781        1
70071        1
70707        1
55887        1
Name: zip_code, Length: 67, dtype: int64

More than half do not have a zip code

In [32]:
ac_new_clean['service_code'].value_counts(dropna=False)

TRAPPED ANIMAL           17814
STRAY                     8632
STRAY AN. SECURED         5497
STRAY AN SECURED          3301
LOOSE                     3009
OWNER SIGNED RELEASE      2570
ABANDONED ANIMAL          2099
INJURED                   1998
CRUELTY                   1697
WILD LIVE                 1350
UNIT STANDING BY          1319
BITE CASE LOOSE            960
ANIMAL RESCUE              905
SICK ANIMAL                795
VET PICKUP                 766
DANGEROUS ANIMAL SIT.      495
BITE CASE SECURE           460
DANGEROUS ANIMAL SIT       409
LOOSE LIVESTOCK            358
SPECIAL CASE               335
SNAKE                      335
SCHOOL CALL                263
DOG FIGHTING               204
UNKNOWN                    182
BITE CASE                  131
OWNED                        4
HIT BY CAR                   1
Name: service_code, dtype: int64

These results are interesting.  This is what the officer reported as the call type which may differ from the request type.  The request type is what the caller reported.  Some categories are duplicated:  
- STRAY AN SECURED/STRAY AN. SECURED
- DANGEROUS ANIMAL SIT/DANGEROUS ANIMAL SIT.

In [33]:
ac_new_clean['species'].value_counts(dropna=False)

DOG          28503
CAT          14508
WILDLIFE     11102
LIVESTOCK      516
FOWL           419
REPTILE        414
BIRD           182
BIRDS          132
OTHER           47
UNKNOWN         33
NaN             27
EXOTIC           6
Name: species, dtype: int64

This data contains incidents related to species other than cats and dogs.  For this analysis we will limit the process to cats and dogs

In [34]:
ac_new_clean['breed'].value_counts(dropna=False)

DSH                   12621
PITBULL                5736
POSSUM                 5412
XPIT BULL              4502
RACCOON                4345
                      ...  
OSTRICH                   1
SAINT BERNARD             1
MOUSE                     1
NORWEGIAN ELKHOUND        1
WOLF                      1
Name: breed, Length: 186, dtype: int64

There are 186 different breeds - There may be too many to use every single one but once cats and dogs are isolated there may be some that can be combined.

In [35]:
ac_new_clean['sex'].value_counts(dropna=False)

M      15217
F      13848
U      13638
NaN    13173
K          4
1          3
D          2
N          2
0          1
S          1
Name: sex, dtype: int64

Many of the unknown sexes may belong to the wild animals.  Will re-evaluate once data set is reduced to cats and dogs.

In [36]:
ac_new_clean['size'].value_counts(dropna=False)

MEDIUM         24385
LARGE          15836
SMALL          12205
NaN             1420
EXTRA LARGE     1313
UNKNOWN          730
Name: size, dtype: int64

No red flags here.  The number of unknown and null are small compared to the size of the dataset.

In [37]:
ac_new_clean['color'].value_counts(dropna=False)

BLK/WHT               4307
BLK                   4301
GRY                   3260
BRWN/WHT              1679
GRY/WHT               1668
                      ... 
WHITE/ BLACK SPOTS       1
BRNDLE                   1
TAMWHT/BLK               1
BLUE DAPPLE              1
WHTGRY                   1
Name: color, Length: 2193, dtype: int64

There are likely too many values here to be valuable in analysis.  Will exclude this column.

In [38]:
ac_new_clean['condition'].value_counts(dropna=False)

FAIR         36625
GOOD          9025
EMACIATED     6253
NaN           2389
UNKNOWN        986
EXCELLENT      307
POOR           304
Name: condition, dtype: int64

This data looks good. There is very little unknown values compared to the size of the dataset.

In [39]:
ac_new_clean['temperment'].value_counts(dropna=False)

NORMAL       25782
FRIENDLY     14132
NERVOUS      12208
NaN           1722
DANGEROUS     1419
UNKNOWN        595
SCARED          31
Name: temperment, dtype: int64

This data looks good. There is very little unknown values compared to the size of the dataset.

In [40]:
ac_new_clean['collar'].value_counts(dropna=False)

NaN           29956
UNKNOWN       16651
NYLON          7087
LEATHER         724
NONE            632
              ...  
CAMO NYLON        1
NYLON/            1
NON               1
NYLON/FLEA        1
N SHOE STR        1
Name: collar, Length: 111, dtype: int64

In [41]:
ac_new_clean['collar'].unique()

array(['NYLON', 'UNKNOWN', 'LEATHER', 'NONE', 'SPIKE', nan, 'ROPE',
       'FLEA', 'HALTERNYLO', 'LEASH', 'PLASTIC', 'BANDANA', 'CHOKE',
       'CLOTH', 'CHAIN', 'SHOCK', 'HARNESS', 'RED STR. J', 'ROPE(PK/YW',
       'NON', 'NYLONN', 'SILVER', 'STRIPES/NY', 'NYLON/FLEA',
       'LEATHER/ST', 'AS 4765', 'COLLAR', 'NYLON STUD', 'LEATH,SPIK',
       'SPIKE LEAT', 'RED/WHT NY', 'NYLON-BONE', 'NYLON PINK', 'HALTER',
       'CAMO', 'WIRE', 'METAL', 'RUBBER', 'PINCH', 'SHIRT', 'CHOK CHAIN',
       'LTHR STUD', 'N', 'NYLON/STUB', 'NLYON', 'STRING', 'NYLON SPIK',
       'LEATHER SP', 'LETHER', 'NEON', 'BLK', 'BRINDLE', 'BELT',
       'NYLON HARN', 'VINYL', 'CHOKER', 'CHAIN/CHRO', 'LSU', 'ELEC', 'NO',
       'NYLON/', 'LARGE CHAI', 'FLEA COLLA', 'LEATHER FL', 'NYLOM',
       'BLUE PINK', 'NYLON/METL', 'NTYLON', 'NYLONCHAIN', 'NYLON/BUCK',
       'CHOKE CHAI', 'BLUE CAMO', 'NYLON FOUN', 'NYLON/DAMA', 'PLEATHER',
       'BRIDLE', 'CHAIN TYPE', 'BARKING', 'nylon', 'SWEATER', 'SPIKED',
       'CAMO 

The type of collar may not be important however the fact that animal was wearing a collar could be a factor that contributes to the result of the animal.  A closer look can also be taken for animals with chains, choke collars can be indicative of cruelty cases.

In [42]:
ac_new_clean['disposition'].value_counts(dropna=False)

TRANS CAA             42528
RETURN TO WILD         9788
RETURN TO OWNER        2137
SPECIAL CASE/OTHER     1000
EUTHANIZED              262
DEAD ON ARRIVAL          84
UNKNOWN                  83
NaN                       7
Name: disposition, dtype: int64

This is the end result of the animal.  This is a possible target variable for a predictive model.  There are very little unknowns and nulls.

In [43]:
ac_new_clean[(ac_new_clean['remarks'].notnull()) & (ac_new_clean['remarks'] !='HOLD')]['remarks']

213                             CAT WAS FOUND IN COMP TRAP
293                       MALE AGGRESSIVE\nFOOD AGGRESSIVE
297            DOG WAS HIT BY A CAR\nBROKEN LEFT FRONT LEG
303      BRITTANY JOHNSON\n504-810-3023 \nINTERESTED IN...
304      BRITTANY JOHNSON\n504-810-3023 \nINTERESTED IN...
                               ...                        
55664                       LOOSE IN YARD AT ABOVE ADDRESS
55702                                    4 PUPPIES WAS DOA
55745        DOG AT LARGE UNDER COMP'S TRUCK\nNO KNOWN OWN
55856            10.7 @ 1231 IMP,TTC\n10.8 @ 1238\nCAGE 10
55883                                            STRAY NKO
Name: remarks, Length: 2504, dtype: object

There is not likely enough information in the remarks to use as a feature.  

### Final data preparation steps

In [44]:
ac_new_clean['service_code'] = ac_new_clean['service_code'].str.replace('.', '')

In [49]:
ac_new_clean['service_code'].unique()

array(['LOOSE', 'TRAPPED ANIMAL', 'STRAY AN SECURED', 'STRAY',
       'OWNER SIGNED RELEASE', 'INJURED', 'SICK ANIMAL', 'WILD LIVE',
       'DANGEROUS ANIMAL SIT', 'BITE CASE LOOSE', 'BITE CASE SECURE',
       'CRUELTY', 'UNIT STANDING BY', 'LOOSE LIVESTOCK', 'VET PICKUP',
       'ANIMAL RESCUE', 'ABANDONED ANIMAL', 'UNKNOWN', 'SPECIAL CASE',
       'SNAKE', 'SCHOOL CALL', 'DOG FIGHTING', 'BITE CASE', 'OWNED',
       'HIT BY CAR'], dtype=object)

In [51]:
ac_new_clean['species'].unique()

array(['DOG', 'WILDLIFE', 'CAT', 'LIVESTOCK', 'BIRDS', 'OTHER', 'EXOTIC',
       'REPTILE', 'FOWL', 'UNKNOWN', nan, 'BIRD'], dtype=object)

In [48]:
ac_new_clean = ac_new_clean.drop(columns=['color','dispatched_situation', 'remarks'])

In [53]:
ac_new_clean.columns

Index(['file_number', 'incident_date', 'incident_time', 'dispatched_time',
       'arrival_time', 'available_time', 'request_type', 'impound_number',
       'impound_date', 'impound_time', 'location_x', 'municipality_',
       'zip_code', 'service_code', 'species', 'breed', 'sex', 'size', 'age',
       'condition', 'temperment', 'collar', 'disposition', 'disposition_date',
       'incident_datetime', 'dispatched_datetime', 'arrival_datetime',
       'available_datetime', 'impound_datetime', 'location_y', 'lat', 'long'],
      dtype='object')

In [54]:
# I want to visualize some data before I drop or change anything else

ac_new_clean.to_csv('../Data/EDA_dataset.csv', index=False)

In [46]:
#to do
#determine if we need to limit to cats and dogs.
#null or unknown dispositions can be eliminated

#questions - are impound numbers unique?  Is the presence or absence of this number meaningful?
#questions - are time to dispatch, time to arrival, time to available potential features?